Make inherited protected internal instance members accessible in class scope.

mawosoft opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This PR enables PowerShell classes to access inherited `protected internal` instance members in addition to `protected` ones. It expands the existing scope checks for `IsFamily` to check for both `IsFamily` and `IsFamilyOrAssembly`.

Fixes #24898.

## PR Context

After switching to .NET 9 in Powershell 7.5, PowerShell classes can no longer call `$this.MemberwiseClone()` because the API signature of the method in the .NET SDK was changed from `protected` to `protected internal`. Other APIs may be affected as well, but this is the most notable problem, causing a **breaking change** when existing scripts switch from PowerShell 7.4 to 7.5 or newer.

To speed up the review process, this PR makes only minimal changes adressing access to inherited **instance** members. These changes should be backported to 7.5 immediately (bugfix). A backport to 7.4 should also be considered (enhancement).

### Future considerations for inherited static members

At the moment, PowerShell classes can only access public inherited static members. If it is desirable to allow access to `protected` and `protected internal` static members as well, I already have a separate PR ready to submit, after this one has been merged.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed: <!-- Number/link of that issue here -->

mawosoft committed:
Add tests for accessing protected (internal) base class members in class scope.

mawosoft committed:
Make protected internal base class instance members accessible in class scope.

mawosoft commented:
<details>
  <summary>Test results <a href="https://github.com/mawosoft/PowerShell/blob/8e3ec8c9b39d170fb98280155dab7139da7a344e/test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1">scripting.Classes.inheritance.tests.ps1</a></summary>

```
[+-] marks success caused by this PR.

Describing Classes inheritance with protected and protected internal members in base class

    Context Derived class can access static base class members
      [-] can access protected base constant field
        PropertyNotFoundException: The property 'Constant' cannot be found on this object. Verify that the property exists.
      [-] can access protected internal base constant field
        PropertyNotFoundException: The property 'Constant' cannot be found on this object. Verify that the property exists.
      [-] can access protected base static field
        PropertyNotFoundException: The property 'StaticField' cannot be found on this object. Verify that the property exists.
      [-] can access protected internal base static field
        PropertyNotFoundException: The property 'StaticField' cannot be found on this object. Verify that the property exists.
      [-] can access protected base static property
        PropertyNotFoundException: The property 'StaticProperty' cannot be found on this object. Verify that the property exists.
      [-] can access protected internal base static property
        PropertyNotFoundException: The property 'StaticProperty' cannot be found on this object. Verify that the property exists.
      [-] can call protected base static method
        RuntimeException: Method invocation failed because [C2Protected] does not contain a method named 'StaticMethod'.
      [-] can call protected internal base static method
        RuntimeException: Method invocation failed because [C2ProtectedInternal] does not contain a method named 'StaticMethod'.

    Context Derived class can access instance base class members
      [+] can call protected base ctor
      [+-] can call protected internal base ctor
      [+] can access protected base instance field
      [+-] can access protected internal base instance field
      [+] can access protected base instance property
      [+-] can access protected internal base instance property
      [+] can call protected base instance method
      [+-] can call protected internal base instance method
      [+] can access protected base virtual property
      [+-] can access protected internal base virtual property
      [+] can call protected base virtual method
      [+-] can call protected internal base virtual method

    Context Derived class can override virtual base class members
      [+] can override protected base virtual property
      [+] can override protected internal base virtual property
      [+] can override protected base virtual method
      [+] can override protected internal base virtual method

    Context Base class members are not accessible outside class scope
      [+] cannot access protected base constant field
      [+] cannot access protected internal base constant field
      [+] cannot access protected base static field
      [+] cannot access protected internal base static field
      [+] cannot access protected base static property
      [+] cannot access protected internal base static property
      [+] cannot call protected base static method
      [+] cannot call protected internal base static method
      [+] cannot access protected base instance field
      [+] cannot access protected internal base instance field
      [+] cannot access protected base instance property
      [+] cannot access protected internal base instance property
      [+] cannot call protected base instance method
      [+] cannot call protected internal base instance method
      [+] cannot access protected base virtual property
      [+] cannot access protected internal base virtual property
      [+] cannot call protected base virtual method
      [+] cannot call protected internal base virtual method
```

</details>

- Tests Passed: 28, Failed: 14 (Baseline)
- Tests Passed: 34, Failed: 8 (All IsFamily tests expandend with IsFamilyOrAssembly)


All `protected internal` base **instance** member types are now accessible in derived PowerShell classes.

As for base **static** member types,  neither `protected internal`  nor `protected` members are accessible in derived PowerShell classes. Doesn't seem like an oversight as with the former case. Not sure if that was a design choice or should be fixed as well?

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1:
```diff
@@ -220,33 +215,28 @@ Describe 'Classes methods with inheritance' -Tags "CI" {
-            {
-                [int]foo() {return 100500}
+            class bar {
+                [int]foo() { return 100500 }
```
@mawosoft Please revert all style changes. If you want you can open new PR for style changes as our guideline requires and we will merge that PR before current one.

mawosoft committed:
Revert accidental style changes as suggested by code review.

iSazonov commented:
@mawosoft Please look CI failures. I see only 8 failed tests.

mawosoft commented:
> @mawosoft Please look CI failures. I see only 8 failed tests.

@iSazonov  The 8 test failures are currently intentional. As mentioned above, they are for access to static members of the base class. This has never worked, even if they are `protected` only.

My questions is: Should this PR be limited to fixing access to instance members of the base class? This is ready and solves the most notable issue with `Object.MemberwiseClone()`.

Access to protected (internal) static members could be adressed in a new PR or not at all if the current behavior is by design.

mawosoft committed:
Refactor tests.
- Remove tests for static members because this PR is limited to instance members.
- Add tests for accessing instance members dynamically.
- Clean up code and clarify test descriptions.

mawosoft commented:
@iSazonov, @SeeminglyScience, @daxian-dbw 
> To speed up the review process, this PR makes only minimal changes adressing access to inherited **instance** members. These changes should be backported to 7.5 immediately (bugfix). A backport to 7.4 should also be considered (enhancement).
> 
> ### Future considerations for inherited static members
> At the moment, PowerShell classes can only access public inherited static members. If it is desirable to allow access to `protected` and `protected internal` static members as well, I already have a separate PR ready to submit, after this one has been merged.

See PR description.

iSazonov commented:
> After switching to .NET 9 in Powershell 7.5, PowerShell classes can no longer call $this.MemberwiseClone()

Can we have a test to ensure this now works?

mawosoft committed:
Add explicit test for $this.MemberwiseClone()

mawosoft commented:
> Can we have a test to ensure this now works?
@iSazonov Done.

daxian-dbw commented:
Sorry for the long delay! I will make sure review this PR next Monday.

daxian-dbw commented on test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1:
```diff
@@ -715,3 +715,209 @@ Describe 'Base type has abstract properties' -Tags "CI" {
+            [C2ProtectedInternal]
+'@
+
+        Add-Type -TypeDefinition $c1DefinitionProtectedInternal
+        Add-Type -TypeDefinition (($c1DefinitionProtectedInternal -creplace 'C1ProtectedInternal', 'C1Protected') -creplace 'protected internal', 'protected')
```
Let's run `Add-Type` only once per process, so as to avoid error when running the tests again.
```suggestion
        if (-not ('C1ProtectedInternal' -as [type])) {
            Add-Type -TypeDefinition $c1DefinitionProtectedInternal
            Add-Type -TypeDefinition (($c1DefinitionProtectedInternal -creplace 'C1ProtectedInternal', 'C1Protected') -creplace 'protected internal', 'protected')
        }
```

mawosoft replied:
Adding the exact same type definition again doesn't produce an error (the definition is hashed if I recall correctly). And if I have made changes to the definition, I don't want a re-run of the tests with the old definition.

daxian-dbw replied:
You are right, I missed that :)

daxian-dbw commented on test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1:
```diff
@@ -715,3 +715,209 @@ Describe 'Base type has abstract properties' -Tags "CI" {
+            }
+            $c3UnrelatedType = Invoke-Expression @"
+                class C3Unrelated {
+                    [void]RunInstanceTest([type]`$derivedType) { $instanceTest }
```
You meant to invoke the script block `$instanceTest`, right?
```suggestion
                    [void]RunInstanceTest([type]`$derivedType) { $instanceTest.Invoke() }
```

mawosoft replied:
No. I want the class method `RunInstanceTest` to have the same inner code as the script block. The actual invocation happens in the `It` block below.

If no class scope is given, the script block is invoked in script scope. Otherwise, the class is instantiated and, by calling the method, the same code runs in a class scope, but not the one needed to successfully access the protected (internal) members.

daxian-dbw replied:
Ah, my bad. This is within an expandable string. You are good.

daxian-dbw reviewed (changes requested):
@mawosoft Again, sorry for the long delay! The product changes look good to me, but I left 2 comments in the test changes. Please take a look, thanks!

daxian-dbw commented:
/azp run PowerShell-CI-linux-packaging PowerShell-Windows-Packaging-CI

mawosoft commented:
@daxian-dbw What are your thoughts on doing this for static members as well? See "future considerations" in the PR Context at the top.

Also, don't forget to backport to 7.5 :) Not sure how that selection process works here.

daxian-dbw commented:
I agree to support the same for static members for the sake of consistency. Please submit a new PR for that. Thank you!
